src/Application.vala

```diff
@@ -296,6 +300,25 @@ public class Terminal.Application : Gtk.Application {
         return 0;
     }
 
+    private void restore_saved_state (Gtk.Window window, bool is_first_window) {
+        if (Application.saved_state.get_boolean ("is-maximized")) {
+            window.maximize ();
+        } else {
+            window.resize (
+                Application.saved_state.get_int ("window-width"),
+                Application.saved_state.get_int ("window-height")
+            );
+        }
+
+        if (is_first_window) {
+            window.size_allocate.connect ((alloc) => {
+                Application.saved_state.set_int ("window-width", alloc.width);
+                Application.saved_state.set_int ("window-height", alloc.height);
```

Comment:
Gotta make sure to skip these if window is maximized otherwise you get weird sizes